consensus/istanbul/backend/handler.go

```diff
@@ -286,3 +301,172 @@ func (sb *Backend) UnregisterPeer(peer consensus.Peer, isProxiedPeer bool) {
 		}
 	}
 }
+
+// Handshake allows the initiating peer to identify itself as a validator
+func (sb *Backend) Handshake(peer consensus.Peer) (bool, error) {
+	// Only written to if there was a non-nil error when sending or receiving
+	errCh := make(chan error)
+	isValidatorCh := make(chan bool)
+
+	sendHandshake := func() {
+		var validatorProofMsg *istanbul.Message
+		var err error
+		peerIsValidator := peer.PurposeIsSet(p2p.ValidatorPurpose)
+		if peerIsValidator {
+			// validatorProofMsg may be nil
+			validatorProofMsg, err = sb.retrieveVersionedEnodeMsg()
+			if err != nil {
+				errCh <- err
+				return
+			}
+		}
+		// Even if we decide not to identify ourselves,
+		// send an empty message to complete the handshake
+		if validatorProofMsg == nil {
+			validatorProofMsg = &istanbul.Message{}
+		}
+		msgBytes, err := validatorProofMsg.Payload()
+		if err != nil {
+			errCh <- err
+			return
+		}
+		err = peer.Send(istanbulValidatorProofMsg, msgBytes)
+		if err != nil {
+			errCh <- err
+		}
+		isValidatorCh <- peerIsValidator
+	}
+	readHandshake := func() {
+		isValidator, err := sb.readValidatorProofMessage(peer)
+		if err != nil {
+			errCh <- err
+			return
+		}
+		isValidatorCh <- isValidator
+	}
+
+	// Only the initating peer sends the message
+	if peer.Inbound() {
+		go readHandshake()
+	} else {
+		go sendHandshake()
+	}
+
+	timeout := time.NewTimer(handshakeTimeout)
+	defer timeout.Stop()
+	select {
+	case err := <-errCh:
+		return false, err
+	case <-timeout.C:
+		return false, p2p.DiscReadTimeout
+	case isValidator := <-isValidatorCh:
+		return isValidator, nil
+	}
+}
+
+// readValidatorProofMessage reads a validator proof message as a part of the
+// istanbul handshake. Returns if the peer is a validator or if an error occurred.
+func (sb *Backend) readValidatorProofMessage(peer consensus.Peer) (bool, error) {
+	logger := sb.logger.New("func", "readValidatorProofMessage")
+	peerMsg, err := peer.ReadMsg()
+	if err != nil {
+		return false, err
+	}
+	if peerMsg.Code != istanbulValidatorProofMsg {
+		logger.Warn("Read incorrect message code", "code", peerMsg.Code)
+		return false, errors.New("Incorrect message code")
+	}
+
+	var payload []byte
+	if err := peerMsg.Decode(&payload); err != nil {
+		return false, err
+	}
+
+	var msg istanbul.Message
+	err = msg.FromPayload(payload, sb.verifyValidatorProofMessage)
+	if err != nil {
+		return false, err
+	}
+	// If the Signature is empty, the peer has decided not to reveal its info
+	if len(msg.Signature) == 0 {
+		return false, nil
+	}
+
+	var versionedEnode versionedEnode
+	err = rlp.DecodeBytes(msg.Msg, &versionedEnode)
+	if err != nil {
+		return false, err
+	}
+
+	node, err := enode.ParseV4(versionedEnode.EnodeURL)
```

Comment:
Should we use `enode.Parse` instead to support the newer `enr` scheme? https://github.com/celo-org/celo-blockchain/blob/5c9f90e6c0495a20473a65c3cd85c516adcfc08f/p2p/enode/node.go#L65

Reply:
My feeling is that we should wait to do this until we decide what our plans are with the ENR scheme and the announce protocol. Maybe this could be left to the new announce protocol changes? It seems like moving to ENR is a step in the right direction and the new announce protocol would be the time to switch